### PR TITLE
Provide a placeholder element for Twitter embeds

### PR DESCRIPTION
**Pull request by westonruter**
Refer to prior work on SoundCloud embeds (https://github.com/ampproject/amp-wp/pull/2722) and YouTube embeds (#3358).

Todo:

- [x] Use Twitter oEmbed response as placeholder content for `amp-twitter`.
- [x] Update tests to mock HTTP requests to prevent external requests during testing.
- [x] Fix tests.
- [x] ~Use YouTube poster image linked to YouTube URL as the `fallback` content for `amp-youtube`.~ See https://github.com/ampproject/amp-wp/pull/1753#issuecomment-511618954. Done in #3358.

# Twitter

Before:

```html
<amp-twitter data-tweetid="1065026506068504577" layout="responsive" width="600" height="480"></amp-twitter>
```

After:

```html
<amp-twitter width="600" height="480" layout="responsive" data-tweetid="1065026506068504577">
    <blockquote class="twitter-tweet" data-width="525" data-dnt="true" placeholder>
        <p lang="en" dir="ltr">🦖 HTM: JSX without the transpiler.<br><br>700 bytes to run in the browser,<br>or
            compile it away using babel-plugin-htm.<br><br>(works with any VDOM library!)<a href="https://t.co/KrlyiYP9Aj">https://t.co/KrlyiYP9Aj</a>
            <a href="https://t.co/qbEFfMT1Dy">pic.twitter.com/qbEFfMT1Dy</a></p>— Jason Miller 🦊⚛ (@_developit) <a
            href="https://twitter.com/_developit/status/1065026506068504577?ref_src=twsrc%5Etfw">November 20, 2018</a>
    </blockquote>
</amp-twitter>
```

# YouTube (Done in #3358)

Before:

```html
<figure class="wp-block-embed-youtube wp-block-embed is-type-video is-provider-youtube  wp-has-aspect-ratio">
    <div class="wp-block-embed__wrapper">
        <amp-youtube data-videoid="8EtmqU844hU" layout="responsive" width="640" height="360"></amp-youtube>
    </div>
</figure>
```

After:

 ```html
<figure class="wp-block-embed-youtube wp-block-embed is-type-video is-provider-youtube  wp-has-aspect-ratio">
    <div class="wp-block-embed__wrapper">
        <amp-youtube data-videoid="8EtmqU844hU" layout="responsive" width="640" height="360">
            <a fallback href="https://www.youtube.com/watch?v=8EtmqU844hU">Finding Portland</a>
            <amp-img placeholder src="https://img.youtube.com/vi/8EtmqU844hU/maxresdefault.jpg" layout="fill"></amp-img>
        </amp-youtube>
    </div>
</figure>
```

### CODE REVIEW

**Comment by westonruter** [comment]
Actually, `fallback` should be used instead of `placeholder`. 

Also, this also needs to be done for `amp-youtube`, though there is not the same amount of content available for use as fallback. For YouTube embedding https://www.youtube.com/watch?v=8EtmqU844hU on a non-AMP page results in the markup:

```html
<figure class="wp-block-embed-youtube wp-block-embed is-type-video is-provider-youtube wp-embed-aspect-16-9 wp-has-aspect-ratio">
    <div class="wp-block-embed__wrapper">
        <iframe title="Finding Portland" width="640" height="360" src="https://www.youtube.com/embed/8EtmqU844hU?feature=oembed" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
    </div>
</figure>
```

And the AMP version is:

```html
<figure class="wp-block-embed-youtube wp-block-embed is-type-video is-provider-youtube  wp-has-aspect-ratio">
    <div class="wp-block-embed__wrapper">
        <amp-youtube data-videoid="8EtmqU844hU" layout="responsive" width="640" height="360"></amp-youtube>
    </div>
</figure>
```

In actuality, the AMP version should have `fallback` content that links to the original URL on YouTube, with the `title` of the `iframe` provided as the link text. In other words:

 ```html
<figure class="wp-block-embed-youtube wp-block-embed is-type-video is-provider-youtube  wp-has-aspect-ratio">
    <div class="wp-block-embed__wrapper">
        <amp-youtube data-videoid="8EtmqU844hU" layout="responsive" width="640" height="360">
            <a fallback href="https://www.youtube.com/watch?v=8EtmqU844hU">Finding Portland</a>
        </amp-youtube>
    </div>
</figure>
```

This will ensure that the linked content is accessible to search engines and also for users who have JS turned off. This should be done generally for all embeds.

For YouTube in particular, a `placeholder` is also appropriate in that we can use the `poster` image as provided by YouTube.

 ```html
<figure class="wp-block-embed-youtube wp-block-embed is-type-video is-provider-youtube  wp-has-aspect-ratio">
    <div class="wp-block-embed__wrapper">
        <amp-youtube data-videoid="8EtmqU844hU" layout="responsive" width="640" height="360">
            <a fallback href="https://www.youtube.com/watch?v=8EtmqU844hU">Finding Portland</a>
            <amp-img placeholder src="https://img.youtube.com/vi/8EtmqU844hU/maxresdefault.jpg" layout="fill"></amp-img>
        </amp-youtube>
    </div>
</figure>
```

**Comment by pierlon** [comment]
@googlebot I consent.

**Comment by pierlon** [comment]
@westonruter should there also be a fallback element for `amp-twitter`?

**Comment by westonruter** [comment]
I don't think so because the `placeholder` has everything. We ended up not using `fallback` for YouTube either.

**Comment by pierlon** [comment]
Ah that's what I thought as well, but the PR title made me think different.

File: includes/embeds/class-amp-twitter-embed-handler.php [hunk 1]
```diff
@@ -46,20 +46,51 @@ class AMP_Twitter_Embed_Handler extends AMP_Base_Embed_Handler {
 	 * Registers embed.
 	 */
 	public function register_embed() {
+		add_filter( 'embed_oembed_html', [ $this, 'filter_embed_oembed_html' ], 10, 2 );
 		add_shortcode( 'tweet', [ $this, 'shortcode' ] ); // Note: This is a Jetpack shortcode.
-		wp_embed_register_handler( 'amp-twitter', self::URL_PATTERN, [ $this, 'oembed' ], -1 );
 		wp_embed_register_handler( 'amp-twitter-timeline', self::URL_PATTERN_TIMELINE, [ $this, 'oembed_timeline' ], -1 );
```
Comment (westonruter) [inline includes/embeds/class-amp-twitter-embed-handler.php]: Eventually the same thing should be done for Twitter timelines.

File: includes/embeds/class-amp-twitter-embed-handler.php [hunk 2]
```diff
@@ -110,12 +141,14 @@ public function shortcode( $attr ) {
...
 	 * Render oEmbed.
 	 *
+	 * @deprecated Since 1.1 as now the sanitize_raw_embeds() is used exclusively, allowing the original oEmbed response to be rapped by amp-twitter.
 	 * @see \WP_Embed::shortcode()
 	 *
 	 * @param array $matches URL pattern matches.
 	 * @return string Rendered oEmbed.
 	 */
 	public function oembed( $matches ) {
+		_deprecated_function( __FUNCTION__, '1.1' );
```
Comment (westonruter) [inline includes/embeds/class-amp-twitter-embed-handler.php]:
```suggestion
		_deprecated_function( __METHOD__, '1.1' );
```

**Comment by csossi** [comment]
Verified in QA

image

**Pull request merged**